ttcn3/types/types.go

```diff
@@ -399,5 +399,35 @@ func printExpr(e syntax.Expr) string {
 }
 
 func TypeOf(n syntax.Expr) Type {
+	//types to check:
+	//ValueLiteral, CompositeLiteral, BinaryExpr, UnaryExpr, Ident
+	switch n := n.(type) {
+	case *syntax.ValueLiteral:
+		switch n.Tok.Kind() {
+		case syntax.INT:
+			return Predefined["integer"]
+		case syntax.FLOAT, syntax.NAN:
+			return Predefined["float"]
+		case syntax.TRUE, syntax.FALSE:
+			return Predefined["boolean"]
+		case syntax.PASS, syntax.FAIL, syntax.INCONC, syntax.NONE, syntax.ERROR:
+			return Predefined["verdicttype"]
+		case syntax.STRING:
+			if strings.IndexFunc(n.Tok.String(), func(r rune) bool {
+				return r <= 32 || r >= 126
+			}) > -1 {
+				return Predefined["universal charstring"]
+			} else {
+				return Predefined["charstring"]
+			}
```

Comment:
I find the flow hard to follow. What about:
```suggestion
			for _, r := range n.Tok.String() {
				if r < 32 || 126 < r {
					return Predefined["universal charstring"]
				}
			}
			return Predefined["charstring"]
```